### PR TITLE
Kokkos: add release 4.1.0

### PR DESCRIPTION
**Pull request by Rombur**

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
Confirmed the version sha256.  Will delay merge to give @janciesko and or @crtrott time to comment.

**Review by masterleinad — APPROVED** [review]

**Comment by eugeneswalker** [comment]
CI failure was due to dependency binary missing from build cache that was expected to be there. @spackbot run pipeline

**Comment by eugeneswalker** [comment]
CI job failed from OneAPI stack:
* [kokkos/sbodnlf 4.1.00 oneapi@=2023.1.0 linux-ubuntu20.04-x86_64 E4S OneAPI](https://gitlab.spack.io/spack/spack/-/jobs/7532467)

Unfortunately the job log was too big to upload so we cannot see what the error was.

**Comment by Rombur** [comment]
@masterleinad do you know of any issue with that compiler?

**Comment by masterleinad** [comment]
> @masterleinad do you know of any issue with that compiler?

No, the error seems to be
```
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-build-sbodnlf/algorithms/unit_tests/sycl/TestSort.cpp:2:
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-src/algorithms/unit_tests/TestSort.hpp:24:
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-src/algorithms/unit_tests/../src/Kokkos_Sort.hpp:70:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/execution:61:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/algorithm_impl.h:31:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/parallel_backend.h:19:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:40:
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:65:21: error: constexpr variable '__mask' must be initialized by a constant expression
    constexpr _UInt __mask = (__is_ascending) ? 1 << ::std::numeric_limits<_Int>::digits : ~_UInt(0) >> 1;
                    ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:200:34: note: in instantiation of function template specialization 'oneapi::dpl::__par_backend_hetero::__order_preserving_cast<true, long long, 0>' requested here
                    auto __val = __order_preserving_cast<__is_ascending>(__val_rng[__val_idx]);
                                 ^
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:614:37: note: in instantiation of function template specialization 'oneapi::dpl::__par_backend_hetero::__radix_sort_count_submit<oneapi::dpl::__par_backend_hetero::__internal::__composite<oneapi::dpl::__par_backend_hetero::__radix_sort_count_kernel<4, true, true>, '_', 'Z', 'T', 'S', 'N', '6', 'o', 'n', 'e', 'a', 'p', 'i', '3', 'd', 'p', 'l', '2', '0', '_', '_', 'p', 'a', 'r', '_', 'b', 'a', 'c', 'k', 'e', 'n', 'd', '_', 'h', 'e', 't', 'e', 'r', 'o', '2', '5', '_', '_', 'r', 'a', 'd', 'i', 'x', '_', 's', 'o', 'r', 't', '_', 'c', 'o', 'u', 'n', 't', '_', 'k', 'e', 'r', 'n', 'e', 'l', 'I', 'L', 'j', '4', 'E', 'L', 'b', '1', 'E', 'L', 'b', '1', 'E', 'J', 'N', 'S', '0', '_', '9', 'e', 'x', 'e', 'c', 'u', 't', 'i', 'o', 'n', '5', '_', '_', 'd', 'p', 'l', '1', '7', 'D', 'e', 'f', 'a', 'u', 'l', 't', 'K', 'e', 'r', 'n', 'e', 'l', 'N', 'a', 'm', 'e', 'E', 'N', 'S', '0', '_', '8', '_', '_', 'r', 'a', 'n', 'g', 'e', 's', '1', '0', 'g', 'u', 'a', 'r', 'd', '_', 'v', 'i', 'e', 'w', 'I', 'P', 'x', 'E', 'E', 'N', '4', 's', 'y', 'c', 'l', '3', '_', 'V', '1', '6', 'b', 'u', 'f', 'f', 'e', 'r', 'I', 'j', 'L', 'i', '1', 'E', 'N', 'S', 'B', '_', '6', 'd', 'e', 't', 'a', 'i', 'l', '1', '7', 'a', 'l', 'i', 'g', 'n', 'e', 'd', '_', 'a', 'l', 'l', 'o', 'c', 'a', 't', 'o', 'r', 'I', 'j', 'E', 'E', 'v', 'E', 'E', 'E', 'E', 'E'>, 4U, true, oneapi::dpl::execution::device_policy<> &, oneapi::dpl::__ranges::guard_view<long long *>, sycl::buffer<unsigned int, 1, sycl::detail::aligned_allocator<unsigned int>, void>, sycl::kernel>' requested here
        sycl::event __count_event = __radix_sort_count_submit<_RadixCountKernel, __radix_bits, __is_ascending>(
                                    ^
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:760:105: note: in instantiation of function template specialization 'oneapi::dpl::__par_backend_hetero::__parallel_radix_sort_iteration<4, true, true>::submit<oneapi::dpl::execution::device_policy<> &, oneapi::dpl::__ranges::guard_view<long long *>, oneapi::dpl::__ranges::all_view<long long, sycl::access::mode::read_write, sycl::access::target::global_buffer, sycl::access::placeholder::true_t> &, sycl::buffer<unsigned int, 1, sycl::detail::aligned_allocator<unsigned int>, void>>' requested here
                __event = __parallel_radix_sort_iteration<__radix_bits, __is_ascending, /*even=*/true>::submit(
                                                                                                        ^
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/glue_algorithm_impl.h:656:30: note: in instantiation of function template specialization 'oneapi::dpl::__internal::__pattern_sort<oneapi::dpl::execution::device_policy<> &, long long *, oneapi::dpl::__internal::__pstl_less>' requested here
    oneapi::dpl::__internal::__pattern_sort(
                             ^
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/glue_algorithm_impl.h:667:18: note: in instantiation of function template specialization 'oneapi::dpl::sort<oneapi::dpl::execution::device_policy<> &, long long *, oneapi::dpl::__internal::__pstl_less>' requested here
    oneapi::dpl::sort(::std::forward<_ExecutionPolicy>(__exec), __first, __last,
                 ^
/tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-src/algorithms/unit_tests/../src/Kokkos_Sort.hpp:694:16: note: in instantiation of function template specialization 'oneapi::dpl::sort<oneapi::dpl::execution::device_policy<> &, long long *>' requested here
  oneapi::dpl::sort(policy, view.data(), view.data() + n);
               ^
/tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-src/algorithms/unit_tests/TestSort.hpp:179:11: note: in instantiation of function template specialization 'Kokkos::sort<long long *, Kokkos::Experimental::SYCL>' requested here
  Kokkos::sort(exec, element_);
          ^
/tmp/root/spack-stage/spack-stage-kokkos-4.1.00-sbodnlffpn5viiacgrvig3deyi5fif2c/spack-src/algorithms/unit_tests/TestSort.hpp:223:13: note: in instantiation of function template specialization 'Test::SortImpl::test_issue_4978_impl<Kokkos::Experimental::SYCL>' requested here
  SortImpl::test_issue_4978_impl<ExecutionSpace>();
            ^
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:65:51: note: shift count 63 >= width of type 'int' (32 bits)
    constexpr _UInt __mask = (__is_ascending) ? 1 << ::std::numeric_limits<_Int>::digits : ~_UInt(0) >> 1;
                                                  ^
``` 
which looks like a compatibility problem between oneDPL and DPC++.

**Comment by eugeneswalker** [comment]
> which looks like a compatibility problem between oneDPL and DPC++.

Do you know @rscohn2 ?

**Comment by masterleinad** [comment]
> Do you know @rscohn2 ?

No, I don't.

**Comment by rscohn2** [comment]
I am using that combination of compiler and dpl and have not seen a problem, although I may not be using sort.

**Comment by Rombur** [comment]
What's the way forward @eugeneswalker ? It works on the machines we have access to.

**Comment by rscohn2** [comment]
How would I reproduce the problem? This works ok for me:
```
(cvenv) rscohn1@anpfclxlin02:spack$ spack install kokkos%oneapi@2023.1.0 +sycl
[+] /usr (external cmake-3.26.3-4qlinkkfibymacsgc6cjbtk26bjf444y)
[+] /usr (external gmake-4.3-dq4khz3mvqcnlrkt2xmt2rj5vvvcvqux)
==> Waiting for kokkos-4.1.00-7za2iku2lq5bxrowin4r74jnsdge3vnp
==> Installing kokkos-4.1.00-7za2iku2lq5bxrowin4r74jnsdge3vnp
==> No binary for kokkos-4.1.00-7za2iku2lq5bxrowin4r74jnsdge3vnp found: installing from source
==> Fetching https://github.com/kokkos/kokkos/archive/4.1.00.tar.gz
==> No patches needed for kokkos
==> kokkos: Executing phase: 'cmake'
==> kokkos: Executing phase: 'build'
==> kokkos: Executing phase: 'install'
==> kokkos: Successfully installed kokkos-4.1.00-7za2iku2lq5bxrowin4r74jnsdge3vnp
  Stage: 3.23s.  Cmake: 1.81s.  Build: 16.30s.  Install: 1.18s.  Post-install: 0.16s.  Total: 22.85s
[+] /localdisk/work/rscohn1/projects/spack/spack/opt/spack/linux-ubuntu22.04-cascadelake/oneapi-2023.1.0/kokkos-4.1.00-7za2iku2lq5bx\
rowin4r74jnsdge3vnp
(cvenv) rscohn1@anpfclxlin02:spack$
```

**Comment by masterleinad** [comment]
> How would I reproduce the problem? This works ok for me:

You would compile with unit tests enabled.

**Comment by rscohn2** [comment]
I can reproduce the problem. Is there a way to pin the oneapi build to the previous version of kokkos until there is a resolution?

**Comment by Rombur** [comment]
Yes, I can do that.

**Comment by rscohn2** [comment]
It is a bug in DPL that was already fixed upstream by https://github.com/oneapi-src/oneDPL/commit/6b876ed8089e319b7fd60ce8694ed84375a57a4b.

We could make a patch for intel-oneapi-dpl, but it looks like kokkos package.py does not have a dependence on dpl. It might be finding dpl because there is an externally installed oneapi and icpx is adding DPL to the include path, but I did not look too closely. To fix the problem, dpl package would have to added to kokkos as a dependence, and then we could make a patch for onedpl.

Since DPL was fixed 2 months ago and there is a release in 1-2 weeks, the next release will have the fix. If we do nothing and try again after the oneapi stack updates to the new release, the error is likely to go away.

I am willing to make the patch, but probably won't if no one is likely to use it because of upcoming release, so let me know what you want to do.

**Comment by Rombur** [comment]
If the problem goes away in a couple of weeks , I am fine waiting.

**Comment by rscohn2** [comment]
Rebased to pick up new DPL in oneapi 2023.2

**Comment by Rombur** [comment]
@masterleinad looks like the oneapi tests are failing because there are too many warnings. There are really only three warnings but we get them every time the file gets included. I am tempted to just disable this warning,  what do you think?

**Comment by masterleinad** [comment]
> @masterleinad looks like the oneapi tests are failing because there are too many warnings. There are really only three warnings but we get them every time the file gets included. I am tempted to just disable this warning, what do you think?

Sure, I don't see much of a reason to care about the warnings in these builds. 
- `-fp-model=precise` would take care of `-Wtautological-constant-compare`
- `-Wno-deprecated-declarations` for the deprecation warnings
- not quite sure if we need to do something about `-Wpass-failed=transform-warning`; there are only two occurrences
- not quite sure if we need to do something about `-Wsycl-target`. This appears once per executable.

**Comment by Rombur** [comment]
@masterleinad  there are more problems with the SYCL backend :-(

```
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-hr3zctgagzpnav2x3febazc5p6czlikn/spack-build-hr3zctg/algorithms/unit_tests/sycl/TestSort.cpp:2:
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-hr3zctgagzpnav2x3febazc5p6czlikn/spack-src/algorithms/unit_tests/TestSort.hpp:24:
In file included from /tmp/root/spack-stage/spack-stage-kokkos-4.1.00-hr3zctgagzpnav2x3febazc5p6czlikn/spack-src/algorithms/unit_tests/../src/Kokkos_Sort.hpp:70:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/execution:61:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/algorithm_impl.h:31:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/parallel_backend.h:19:
In file included from /opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:40:
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:65:21: error: constexpr variable '__mask' must be initialized by a constant expression
    constexpr _UInt __mask = (__is_ascending) ? 1 << ::std::numeric_limits<_Int>::digits : ~_UInt(0) >> 1;
                    ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/intel/oneapi/dpl/2022.1.0/linux/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h:200:34: note: in instantiation of function template specialization 'oneapi::dpl::__par_backend_hetero::__order_preserving_cast<true, long long, 0>' requested here
                    auto __val = __order_preserving_cast<__is_ascending>(__val_rng[__val_idx]);
                                 ^
```

**Comment by masterleinad** [comment]
> @masterleinad there are more problems with the SYCL backend :-(

That looks like the same `oneDPL` we have seen before (https://github.com/spack/spack/pull/38641#issuecomment-1617120806). Looking at https://github.com/spack/spack/pull/38641#issuecomment-1640014607, are we using `one API2023.2` now (which should pick up a newer `oneDPL` as well)?

File: var/spack/repos/builtin/packages/kokkos/package.py [hunk 1]
```diff
@@ -329,6 +330,9 @@ def cmake_args(self):
                 self.define("CMAKE_CXX_COMPILER", self.spec["kokkos-nvcc-wrapper"].kokkos_cxx)
             )
 
+        if "+sycl" in self.spec:
```
Comment (rscohn2) [inline var/spack/repos/builtin/packages/kokkos/package.py]:
```suggestion
        if self.spec.satisfies("%oneapi") or self.spec.satisfies("%intel"):
```

**Comment by rscohn2** [comment]
This builds correctly using your branch
```
spack install kokkos%oneapi@2023.2.0 +sycl +tests
```
The issue may be that the oneapi software stack that is tested in gitlab needs to be updated to 2023.2. @eugeneswalker would know. If that is not going to happen soon, I will do a patch for the 2023.1 DPL.

**Comment by eugeneswalker** [comment]
> The issue may be that the oneapi software stack that is tested in gitlab needs to be updated to 2023.2. @eugeneswalker would know. If that is not going to happen soon, I will do a patch for the 2023.1 DPL.

Yes, our OneAPI CI container is based on prior 2023.1.0 release. I am upgrading the container now based on 2023.2, can submit a PR today that will make that the one used for Spack CI today, may take a few days to get merged. Once merged, we could rebase this PR on develop and it would be building with the latest.

Edit: PR here:
* https://github.com/spack/spack/pull/39048

**Comment by tldahlgren** [comment]
> > The issue may be that the oneapi software stack that is tested in gitlab needs to be updated to 2023.2. @eugeneswalker would know. If that is not going to happen soon, I will do a patch for the 2023.1 DPL.
> 
> Yes, our OneAPI CI container is based on prior 2023.1.0 release. I am upgrading the container now based on 2023.2, can submit a PR today that will make that the one used for Spack CI today, may take a few days to get merged. Once merged, we could rebase this PR on develop and it would be building with the latest.
> 
> Edit: PR here:
> 
> * [e4s oneapi ci: build with latest 2023.2 based image #39048](https://github.com/spack/spack/pull/39048)

I noticed the PR was merged so clicked `Update branch`.

**Review by tldahlgren — APPROVED** [review]

**Comment by balay** [comment]
any reason why `kokkos-kernels@4.1.0` was not added here?

**Comment by masterleinad** [comment]
> any reason why `kokkos-kernels@4.1.0` was not added here?

Kokkos and KokkosKernels are different packages. 🤷

**Comment by balay** [comment]
Ok.

 I'm seeing some build issues with it [will followup on the 4.1.00 PR] - but build of 4.0.00 goes through fine - so was checking if that was the reason. 

And then `trilinos@14.4.0` has a dependency on `kokkos@4.1.00`  - that means `kokkos-kernels@4.1.00` needs to be pulled in [if kokkos-kernels is also needed] - hence this issue..

**Pull request merged**